File: equinox-web-csharp/Domain/Aggregate.fs
module TodoBackendTemplate.Aggregate

// NB - these types and names reflect the actual storage formats and hence need to be versioned with care
module Events =
    type Compacted = { happened: bool }

    type Event =
        | Happened
        | Compacted of Compacted
        interface TypeShape.UnionContract.IUnionContract

module Folds =

    type State = { happened: bool }
    let initial = { happened = false }
    let evolve s = function
        | Events.Happened -> { happened = true }
        | Events.Compacted e -> { happened = e.happened} 
    let fold (state : State) : Events.Event seq -> State = Seq.fold evolve state
    let isOrigin = function Events.Compacted _ -> true | _ -> false
    let compact state = Events.Compacted { happened = state.happened }

module Commands =

    type Command =
        | MakeItSo

    let interpret c (state : Folds.State) =
        match c with
        | MakeItSo -> if state.happened then [] else [Events.Happened]

type Handler(log, stream, ?maxAttempts) =

    let inner = Equinox.Handler(Folds.fold, log, stream, maxAttempts = defaultArg maxAttempts 2)

    /// Execute `command`, syncing any events decided upon
    member __.Execute command : Async<unit> =
        inner.Decide <| fun ctx ->
            ctx.Execute (Commands.interpret command)
    /// Establish the present state of the Stream, project from that as specified by `projection`
    member __.Query(projection : Folds.State -> 't) : Async<'t> =
        inner.Query projection

type View = { sorted : bool }

type Service(handlerLog, resolve) =
    
    let (|CategoryId|) (id: string) = Equinox.CatId("Aggregate", id)
    
    /// Maps a ClientId to Handler for the relevant stream
    let (|Stream|) (CategoryId catId) = Handler(handlerLog, resolve catId)

    let render (s: Folds.State) : View =
        {   sorted = s.happened }

    /// Read the present state
    // TOCONSIDER: you should probably be separating this out per CQRS and reading from a denormalized/cached set of projections
    member __.Read(Stream stream) : Async<View> =
        stream.Query (fun s -> render s)

    /// Execute the specified command 
    member __.Execute(Stream stream, command) : Async<unit> =
        stream.Execute command

File: equinox-web-csharp/Domain/Aggregate.cs
using Microsoft.FSharp.Core;

namespace TodoBackendTemplate;

public static class Aggregate
{
    public interface IEvent { }
    /// NB - these types and names reflect the actual storage formats and hence need to be versioned with care
    public static class Event
    {
        public record Happened : IEvent { }
        public record Snapshotted(bool HasHappened) : IEvent;

        static readonly SystemTextJsonUtf8Codec Codec = new(new System.Text.Json.JsonSerializerOptions());

        public static IEvent? TryDecode(string et, ReadOnlyMemory<byte> json) =>
            et switch
            {
                nameof(Happened) => Codec.Decode<Happened>(json),
                nameof(Snapshotted) => Codec.Decode<Snapshotted>(json),
                _ => null
            };

        public static (string, ReadOnlyMemory<byte>) Encode(IEvent e) => (e.GetType().Name, Codec.Encode(e));
        public const string Category = "Aggregate";
        public static string StreamId(ClientId id) => id.ToString();
    }

    public class State
    {
        public bool Happened { get; set; }

        State(bool happened) { Happened = happened; }

        public static readonly State Initial = new(false);

        static void Evolve(State s, IEvent x) =>
            s.Happened = x switch
            {
                Event.Happened => true,
                Event.Snapshotted e => e.HasHappened,
                _ => throw new ArgumentOutOfRangeException(nameof(x), x, "invalid")
            };

        public static State Fold(State origin, IEnumerable<IEvent> xs)
        {
            // NB Fold must not mutate the origin
            var s = new State(origin.Happened);
            foreach (var x in xs)
                Evolve(s, x);
            return s;
        }

        public static bool IsOrigin(IEvent e) => e is Event.Snapshotted;

        public static IEvent Snapshot(State s) => new Event.Snapshotted(s.Happened);
    }

    /// Defines the decision process which maps from the intent of the `Command` to the `Event`s that represent that decision in the Stream
    public abstract class Command
    {
        public class MakeItSo : Command { }

        public IEvent[] Interpret(State s) => this switch
        {
            MakeItSo => s.Happened ? [] : [new Event.Happened()],
            _ => throw new ArgumentOutOfRangeException(nameof(Command), this, "invalid")
        };
    }

    public record View(bool Sorted);

    public class Service(Func<ClientId, Equinox.DeciderCore<IEvent, State>> resolve)
    {
        /// Execute the specified command
        public Task<Unit> Execute(ClientId id, Command command) =>
            resolve(id).Transact(command.Interpret);

        /// Read the present state
        // TOCONSIDER: you should probably be separating this out per CQRS and reading from a denormalized/cached set of projections
        public Task<View> Read(ClientId id) =>
            resolve(id).Query(Render);

        static View Render(State s) => new(Sorted: s.Happened);
    }
}